sonic-xcvrd/xcvrd/xcvrd.py

```diff
@@ -1000,12 +1000,22 @@ def is_cmis_application_update_required(self, api, app_new, host_lanes_mask):
         for lane in range(self.CMIS_MAX_HOST_LANES):
             if ((1 << lane) & host_lanes_mask) == 0:
                 continue
+            app_cur = api.get_application(lane)
             if app_old == 0:
-                app_old = api.get_application(lane)
-            elif app_old != api.get_application(lane):
+                app_old = app_cur
+            elif app_cur != 0:
+                if app_cur != app_new:
+                    self.log_notice("Changing from default AppSel {} to non default "
+                                    "AppSel code {}. Reset AppSel code for all lanes"
+                                     .format(app_cur, app_new))
+                    api.set_application(0xff, 0, 0)
+                    api.set_datapath_deinit(0xff)
+                    api.scs_apply_datapath_init(0xff)
+                    return True
```

Comment:
@AnoopKamath The function name does not indicate changing the module state as done by these lines. This looks to be a hacky approach

Reply:
@AnoopKamath Can you please create a separate function (clear_cmis_app_code_if_needed or some meaningful name) to clear the application. Also, you can call this function before calling is_cmis_application_update_required to have separate functions to achieve different tasks.

Also, the current check (line 1001-1002) to skip handling lanes which are not part of current host lanes can cause to skip clearing of default app code value in the following case.
1) Let's say default appl_code for lanes 1-4 = 1 and appl_code for lanes 5-8 = 0
2) Now, CMIS init is requested for a subport to set lanes 5-6 with appl_code = 3 (goal is to have all lanes 1-2, 3-4, 5-6 and 7-8 to have appl_code = 3 eventually for the corresponding physical port). However, the current function will not clear lanes 1-4 since CMIS init is triggered for lanes 5-6. This would cause an issue since the current state of the lanes would be invalid with lanes 1-4 with appl_code = 1, lanes 5-6 with appl_code = 3 and lanes 7-8 with appl_code = 0.
In order to prevent such issue, can we remove the host_lanes_mask check in the new function? We can have the new function similar to below. Please feel free to let me know if you find any issues with this implementation.
```
def clear_cmis_app_code_if_needed(api, app_new):
    for lane in range(self.CMIS_MAX_HOST_LANES):
    app_cur = api.get_application(lane)
    if app_cur != 0 and app_cur != app_new:
	    clear_app_code_for_all_lanes(api)
	    break
```

Reply:
Taken care